### PR TITLE
add wlr_output_layout_adjacent_output

### PR DESCRIPTION
**Pull request by acrisci**
This function is for determining the adjacent output to another output, like we have to do with sway in some places.

The output extents must be entirely within the direction from the corresponding border of the reference output to be considered. This function then picks the closest output to the reference point.

TODO:

* [x] test the function

### CODE REVIEW

**Comment by acrisci** [comment]
Here is an important edge case:

```
+-----------------+
|                 |
|                 |
|       a         |
|                 |
|                 |
+-----------------+

            +-------------------+
            |                   |
            |                   |
            |        b          |
            |                   |
            |                   |
            +-------------------+
```

In this case, `a` is *not* left of `b` because it contains a region that is not to the left of the left border of `b`.

**Comment by SirCmpwn** [comment]
Also test the case of mirrored and partially mirrored outputs

File: include/wlr/util/direction.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#ifndef WLR_UTIL_DIRECTION_H
+#define WLR_UTIL_DIRECTION_H
+
+enum wlr_direction {
+	WLR_DIRECTION_UP = 0,
+	WLR_DIRECTION_DOWN = 1,
+	WLR_DIRECTION_LEFT = 2,
+	WLR_DIRECTION_RIGHT = 4,
+};
```
Comment (SirCmpwn) [inline include/wlr/util/direction.h]: I don't think this merits its own header

File: types/wlr_output_layout.c [hunk 2]
```diff
@@ -421,3 +421,53 @@ struct wlr_output *wlr_output_layout_get_center_output(
 
 	return wlr_output_layout_output_at(layout, dest_x, dest_y);
 }
+
+
+struct wlr_output *wlr_output_layout_adjacent_output(
+		struct wlr_output_layout *layout, enum wlr_direction direction,
+		struct wlr_output *reference, double ref_x, double ref_y) {
+	// XXX should we allow reference to be NULL?
```
Comment (SirCmpwn) [inline types/wlr_output_layout.c]: Why?

File: include/wlr/util/direction.h (hunk 1 above)
Comment (acrisci) [reply include/wlr/util/direction.h]: Where do you want it?

File: types/wlr_output_layout.c (hunk 2 above)
Comment (acrisci) [reply types/wlr_output_layout.c]: The only reason I can think of is to make it consistent with the rest of the api. The rule generally is that if reference is not passed, then it acts on the whole layout, but there's no concrete use case for this yet.

File: include/wlr/util/direction.h (hunk 1 above)
Comment (SirCmpwn) [reply include/wlr/util/direction.h]: Can just live in wlr_output_layout.h imo

File: types/wlr_output_layout.c (hunk 2 above)
Comment (SirCmpwn) [reply types/wlr_output_layout.c]: I don't know how that could work, probably best not to.

**Comment by acrisci** [comment]
This is now used in [sway#1585](https://github.com/swaywm/sway/pull/1585).

File: types/wlr_output_layout.c (hunk 2 above)
Comment (SirCmpwn) [reply types/wlr_output_layout.c]: Should drop the comment, leave the assert

Comment (emersion) [reply types/wlr_output_layout.c]: Can't we automatically get the `reference` output from `ref_x` and `ref_y`?

Comment (acrisci) [reply types/wlr_output_layout.c]: We could, but we need to start being more careful about supporting the overlapping output case. There are a few functions that currently don't support this well that need to be revisited.
